src/blockchain/contracts/TwoPhaseCommit.sol

```diff
@@ -43,33 +50,83 @@ contract TwoPhaseCommit {
         uint32 cohort_id,
         Ballot ballot
     ) public {
+        require(
+            getNow() < transaction_configs[transaction_id].vote_timeout_time
+        );
+        require(transaction_configs[transaction_id].cohorts > cohort_id);
+        require(ballot != Ballot.UNKNOWN);
         transaction_states[transaction_id][cohort_id] = ballot;
     }
 
-    enum VotingDecision {
+    enum VotingDecisionOption {
         UNKNOWN, // Saved for unexpected unset value only.
         PENDING,
         COMMIT,
         ABORT
     }
 
+    struct VotingDecision {
+        VotingDecisionOption option;
+        string reason;
+    }
+
     // Gets the voting decision of a transaction.
     function getVotingDecision(string memory transaction_id)
```

Comment:
IIUC, this function will rely on the coordinator to deliver the message to blockchain before it times out, right?

Reply:
I think each cohort will send its own vote to the blockchain. @heronyang Is that your understanding as well?

Reply:
@n-triple-a I'm unsure what you mean by 'delivering the message to blockchain'. The flow is: (1) coordinator calls `startVoting()`, (2) each cohort calls `vote()`, then (3) anyone can call `getVotingDecision()` to get the result after times out.

@benjmarks22 Correct.

Reply:
Thanks Heron, this makes sense to me.